torch/utils/data/dataloader_experimental.py

```diff
@@ -87,8 +89,11 @@ def __new__(cls,
                 raise Exception(
                     'sampler is not yet supported by DataPipes')
             datapipe = dataset
-            if shuffle:
-                # Enforce at least one shuffle in the graph
+            # Enforce at least one shuffle in the graph if shuffle=True
+            if shuffle and not any(
+                    isinstance(dp, Shuffler)
+                    for dp in torch.utils.data.graph_settings.get_all_graph_pipes(traverse(dataset, only_datapipe=True))
+            ):
```

Comment:
1. We should probably warn if we need to add a `Shuffler`, since the default buffer size may not be sufficient for the task at hand. Otherwise the user might see significantly degraded performance without any indication.
2. It would merge this into 

    https://github.com/pytorch/pytorch/blob/fa1a41ca718fd45a29ac1655861f40703c875c1b/torch/utils/data/graph_settings.py#L30

    given that it is already looking for `Shuffler`'s. Not sure about BC though. Is this considered public?

Reply:
1. I totally agree
2. I think it's fine. This is not public and we even have a plan to bring DataLoader2 into TorchData and spin off TorchData from PyTorch